src/DPORDriver.cpp

```diff
@@ -192,17 +195,209 @@ Trace *DPORDriver::run_once(TraceBuilder &TB, bool &assume_blocked) const{
+      print_progress(computation_count, estimate, res);
+    }
+
+    bool assume_blocked = false;
+    TB->reset();
+    Trace *t= this->run_once(*TB, assume_blocked);
+    tasks_left--;
+
+    int to_create = TB->tasks_created;
+
+
+    tasks_left += to_create;
+
+    if (handle_trace(TB, t, &computation_count, res, assume_blocked)) {
+      break;
+    }
+    if(conf.print_progress_estimate && (computation_count+1) % 100 == 0){
+      estimate = std::round(TB->estimate_trace_count());
+    }
+
+  } while(tasks_left);
+
+  if(conf.print_progress){
+    llvm::dbgs() << ESC_char << "[K\n";
+  }
+
+  SigSegvHandler::reset_signal_handler();
+
+  delete TB;
+
+  return res;
+}
+
+DPORDriver::Result DPORDriver::run_rfsc_parallel() {
+
+  Result res;
+
+  int n_threadrunners = conf.n_threads-1;
+
+  moodycamel::BlockingConcurrentQueue<std::tuple<Trace *, bool, int>> queue;
+  std::tuple<Trace *, bool, int> tup;
+
+  RFSCDecisionTree decision_tree;
+  RFSCUnfoldingTree unfolding_tree;
+
+  ctpl::thread_pool threadpool(n_threadrunners);
+
+  std::vector<RFSCTraceBuilder*> TBs;
+  for (int i = 0; i < n_threadrunners; i++) {
+    TBs.push_back(new RFSCTraceBuilder(decision_tree, unfolding_tree, conf));
+  }
+
+  auto thread_runner = [this, &TBs, &queue] (int id) {
+    bool assume_blocked = false;
+    if (TBs[id]->reset()) {
+      Trace *t= this->run_once(*TBs[id], assume_blocked);
+      // Release shared_ptr when finished
+      TBs[id]->work_item = nullptr;
+
+      queue.enqueue(std::make_tuple(std::move(t), assume_blocked, TBs[id]->tasks_created));
+    }
+  };
+
+  SigSegvHandler::setup_signal_handler();
+
+
+  uint64_t computation_count = 0;
+  long double estimate = 1;
+
+  // Initialize the first run.
+  threadpool.push(thread_runner);
+  int tasks_left = 1;
+
+  do{
+    if(conf.print_progress){
+      print_progress(computation_count, estimate, res);
+    }
+
+    queue.wait_dequeue(tup);
+    tasks_left--;
+
+    Trace *t;
+    bool assume_blocked;
+    int to_create;
+    std::tie(t, assume_blocked, to_create) = tup;
+
+    tasks_left += to_create;
+
+    // handle_trace requires a TB but with RFSC this is a constant value, so it does not matter which TB it is.
+    if (handle_trace(TBs[0], t, &computation_count, res, assume_blocked)) {
+      tasks_left = 0;
+      threadpool.stop(false);
+      break;
+    }
+    // TODO: Estimations are run on a TB prefix, for now we ignore this functionality.
+    // if(conf.print_progress_estimate && (computation_count+1) % 100 == 0){
+    //   estimate = std::round(TB->estimate_trace_count());
+    // }
+
+    for(int i = 0; i < to_create; i++) {
+      threadpool.push(thread_runner);
+    }
+
+  } while(tasks_left);
+
+  if(conf.print_progress){
+    llvm::dbgs() << ESC_char << "[K\n";
+  }
+
+  SigSegvHandler::reset_signal_handler();
+
+  // Spinlock to make sure all worker threads have finished their task before deleting TraceBuilders
+  while (threadpool.n_idle() != n_threadrunners);
```

Comment:
You are correct in that it only blocks on a premature stop. I chose to have it just above the deletion to try and make it clear the reason for why the spinlock exist, not when it is required.
But it could be moved up to `stop(false)`.

Reply:
I think that's more clear to have it higher up, but I also asked because when first reading I reacted to seeing a busy-wait loop. Knowing that it, at least, is only used for the error case, makes it a little less urgent to fix.